frontend/actions.js

```diff
@@ -98,58 +89,149 @@ export default ({setState, getState}) => {
+    sendAddress: Object.assign(state.sendAddress, {validationMsg: ''}),
   })
 
-  // is being called debounced, thus we need to getState
-  // TODO maybe it'd be good practice to only use getState and avoid this problem?
+  const validateSendAddress = (state) => {
+    if (!Cardano.isValidAddress(state.sendAddress.fieldValue)) {
+      setState({
+        sendAddress: Object.assign(state.sendAddress, {validationMsg: 'Invalid address'}),
+      })
+      return false
+    }
+    return true
+  }
+
+  const validateSendAmount = (state) => {
+    if (isNaN(parseFloat(state.sendAmount.fieldValue))) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: 'Invalid format: Amount has to be a number',
+        }),
+      })
+      return false
+    }
+    if (parseFloat(state.sendAmount.fieldValue).toFixed(6) <= 0) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: 'Invalid format: Amount has to be a positive number',
+        }),
+      })
+      return false
+    }
+    if (parseFloat(state.sendAmount.fieldValue).toFixed(6) > state.balance / 1000000) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: `Insufficient funds. Your balance is ${state.balance / 1000000} ADA.`,
+        }),
+      })
+      return false
+    }
+    return true
+  }
+
+  const validateFee = (state) => {
+    if (parseFloat(state.transactionFee) < 0) {
+      setState({
+        transactionFee: -state.transactionFee,
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: `Insufficient funds to cover the transaction fee. Send at most ${(state.balance +
+            state.transactionFee) /
+            1000000} ADA.`,
+        }),
+      })
+      return false
+    }
+    return true
+  }
+
+  const validateSendForm = (state) => {
+    if (state.sendAddress.fieldValue !== '' && state.sendAmount.fieldValue !== '') {
+      clearSendAddressValidation(state)
+      clearSendAmountValidation(state)
+      const isValidAddress = validateSendAddress(state)
+      const isValidAmount = validateSendAmount(state)
+      return isValidAddress && isValidAmount
+    }
+    return false
+  }
+
   const calculateFee = async () => {
     const state = getState()
-    const address = state.sendAddress
-    const sendValue = state.sendAmountFieldValue
-    const amount = parseFloat(state.sendAmountFieldValue || 0)
-    const transactionFee = await wallet.getTxFee(address, amount)
-    // if we reverted value in the meanwhile, do nothing, otherwise update
-    if (state.sendAmountFieldValue !== state.feeCalculatedFrom) {
-      setState({transactionFee, feeCalculatedFrom: sendValue})
+    if (state.sendAddress.fieldValue !== '' && state.sendAmount.fieldValue !== '') {
+      setState({calculatingFee: true})
+      // if we reverted value in the meanwhile, do nothing, otherwise update
+      if (
```

Comment:
this check does not make sense where it is - you've only made sync calls so far

you need to check for this after calling `await wallet.getTxFee(address, amount)` and before setting state

Reply:
here it prevents unnecessary fee calculations, but understood